### DISCUSSION TITLE
None guard with boolean variable with other conditions.

### DISCUSSION THREAD

**Discussion by diego-pm** [microsoft/pyright#4032] [Q&A] [answered]
I am trying to create a variable that indicates if another variable is not None, but also checking some other conditions, such as if a list is not empty. However, when adding the other conditions, pyright does not detect the None guard.

For example, pyright does not detect any errors in this case (detects the none guard):

```python
def f(a: list[int] | None) -> int:
  not_none = a is not None
  if not_none:
      return a[0]
  return 0
```

But it gives an error in the following case (does not detect the none guard):

```python
def f(a: list[int] | None) -> int:
  not_none = a is not None and len(a) > 0
  if not_none:
      return a[0]
  return 0
```

How could I fix the second case? The guarding would work well without the `not_none` variable and using the boolean expression directly in the if clause, but I really want to use the intermediate variable `not_none` in order to reuse the vale of the statement.

**Comment by erictraut** [comment, marked as answer]
Normally, you need to include the narrowed expression within the conditional check itself. The reason the top sample works is because pyright supports an advanced feature called [Aliased Conditional Expressions](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#aliased-conditional-expression). This feature was recently added to the TypeScript compiler, and I borrowed the idea for pyright. Note that the top code sample will not type check with other Python type checkers like mypy because they don't have support for Aliased Conditional Expressions (although it is a frequently-requested feature for mypy, so it may be added some day).

There are limits to Aliased Conditional Expressions. It works only with certain types of expressions, and `and` and `or` operators are not allowed. Your second code sample uses an expression form that is too complex, so the feature no longer works in this situation, and type narrowing for `a` is not applied.

If you want the second code sample to type check without errors, you'll need to make modifications to it. Here are a few alternatives that work:

```python
def f(a: list[int] | None) -> int:
  not_none = a is not None
  if not_none and len(a) > 0:
      return a[0]
  return 0
```

```python
from typing import TypeGuard

def is_nonempty_list(a: list[int] | None) -> TypeGuard[list[int]]:
    return a is not None and len(a) > 0

def f(a: list[int] | None) -> int:
  if is_nonempty_list(a):
      return a[0]
  return 0
```

**Reply by diego-pm** [reply]
Thank you for your suggestions and explanations! Are there any plan to support more complex expressions in Aliased Conditional Expressions?
